### PR TITLE
Online draft autosave and Imported location parity

### PR DESCRIPTION
**Pull request by glittle**
grokbot: Local #303 UAT fixes Glen already tested.

## Summary
- Online ballot **Draft** autosave (OnlineBallotStatus.Draft / SubmitOnlineBallotDto.IsDraft) so in-progress votes stay off Accept-all pending until real submit
- **Imported** location parity with Online: reserved type, i18n label, sort-only edit, no teller-created ballots
- Supporting FE (voter ballot page, computer codes, locations UI/tests) and context docs

## Test plan
- [ ] Online voter: edit ballot, leave/return — draft restores; Accept-all does not pull drafts
- [ ] Submit draft → appears in pending; Accept-all processes it
- [ ] Locations: Imported shows reserved label; cannot rename/edit like Online; teller cannot start ballot at Imported
- [ ] Existing Online reserved behavior still correct

### CODE REVIEW

**Review by glittle — COMMENTED** [review]
grokbot: Requesting Copilot-style review (in progress).

File: frontend/src/pages/voting/VoterBallotPage.vue [hunk 1]
```diff
@@ -217,6 +269,23 @@ async function handleSubmit() {
...
 
+watch(
+  () =>
+    votes.value.map((v) => ({
+      personGuid: v.person?.personGuid,
+      searchText: v.searchText,
+      freeText: v.freeText,
+    })),
+  () => {
+    runAutosave();
```
Comment (glittle) [inline frontend/src/pages/voting/VoterBallotPage.vue]: grokbot: **Blocks merge — autosave always sends `isDraft: true`.**

After load, `applyPriorVotes` fills the slots, then `autosaveReady` flips true and this deep `watch` (plus the notify watcher / input handlers) calls `runAutosave` → `buildSubmitPayload(true)`.

For a voter who already **Submitted**, that silent call demotes the row to Draft within ~800ms with no further user action (see companion comment on `TryWritePendingPayloadIfEditableAsync`). Same path after an explicit Submit while they keep editing.

Even with a server-side no-demote guard, autosave for an already-submitted editable ballot should keep `isDraft: false` (or skip status changes and only patch payload) so client and server agree.

File: backend/Services/OnlineVotingService.Ballot.cs [hunk 2]
```diff
@@ -74,11 +74,14 @@ public partial class OnlineVotingService
...
+            var targetStatus = dto.IsDraft
+                ? OnlineBallotStatus.Draft
+                : OnlineBallotStatus.Submitted;
 
             if (existingVotingInfo != null)
             {
-                var wrote = await TryWritePendingPayloadIfStillSubmittedAsync(
-                    existingVotingInfo, payloadJson, now);
+                var wrote = await TryWritePendingPayloadIfEditableAsync(
+                    existingVotingInfo, payloadJson, now, targetStatus);
```
Comment (glittle) [inline backend/Services/OnlineVotingService.Ballot.cs]: grokbot: **Blocks merge — Submitted can be demoted to Draft.**

`targetStatus` follows `dto.IsDraft` even when the existing row is already `Submitted`. `TryWritePendingPayloadIfEditableAsync` allows writes while Status is Draft or Submitted, then sets Status to `targetStatus`.

So any `IsDraft: true` call on a Submitted row (silent autosave after reload, after an explicit Submit while still editing, or toggling notify) writes Status back to `Draft` and drops the ballot out of Accept-all pending / monitor Submitted counts — the opposite of the Draft-vs-Submitted contract in `context/online-ballots.md`.

Fix on the server (do not trust the client alone): never demote. If the row is already Submitted, keep Submitted even when `IsDraft` is true (payload update only). Prefer Draft to Draft, Draft to Submitted, Submitted to Submitted only.

Please add a unit test: Submitted + `IsDraft: true` stays Submitted and stays Accept-all pending.

**Review by glittle — COMMENTED** [review]
grokbot: Copilot-style review of #321 — **not ready**.

**Blocks merge**
1. Backend: `IsDraft: true` can demote an existing `Submitted` row to `Draft` (drops Accept-all / monitor pending).
2. Frontend: silent autosave always sends `isDraft: true`; after reload of a Submitted ballot the deep watch fires it within ~800ms with no further user action.

**Also (non-blocking)**
- No new test that Draft is excluded from Accept-all, or that Submitted cannot demote.
- `Person.HasOnlineBallot` is now set on Draft autosave; `context/people.md` still describes it as set on Submitted submit — update that doc when fixing.

Imported location parity / display helpers look consistent with Online (type-based, not name). Converted back to draft until the demotion path is fixed.

File: frontend/src/pages/voting/VoterBallotPage.vue [hunk 3]
```diff
@@ -121,16 +130,61 @@ async function loadElectionData() {
...
+    electionGuid: electionGuid.value,
+    voterId: onlineVotingStore.voterId!,
+    votes: onlineVotes,
+    listPool: poolEntries.value,
+    notifyWhenProcessed: notifyWhenProcessed.value,
+    isDraft,
+  };
+}
+
+const runAutosave = debounce(async () => {
```
[the remark this answers is not in the document]
Comment (glittle) [reply frontend/src/pages/voting/VoterBallotPage.vue]: grokbot: Confirmed on `32ab223` — `StatusAfterWrite` / ExecuteUpdate never demote Submitted, FE sends `isDraft: false` once Submitted, and tests cover Draft excluded from Accept-all plus Submitted+IsDraft stays pending. Resolving this thread.

Remaining: cancel/flush of the debounce still helps avoid orphan in-flight requests after navigate, but it no longer demotes Status. Not blocking on demotion anymore.

File: frontend/src/pages/voting/VoterBallotPage.vue [hunk 4]
```diff
@@ -121,16 +130,61 @@ async function loadElectionData() {
...
+  if (!autosaveReady.value || !canChangeVote.value || submitting.value) {
+    return;
+  }
+  if (!hasAnyVote(votes.value)) {
+    return;
+  }
```
[the remark this answers is not in the document]
Comment (glittle) [reply frontend/src/pages/voting/VoterBallotPage.vue]: grokbot: Still blocks merge on `32ab223`. `if (!hasAnyVote(votes.value)) return` still skips overwrite, so clearing the last name leaves the prior Draft payload on the server and a leave/return restores the names the voter just removed. Empty drafts need to overwrite (or clear) the saved payload.

File: frontend/src/composables/useVoterBallot.ts [hunk 5]
```diff
@@ -166,20 +182,59 @@ export function useVoterBallotHelpers(selectionMode: () => string) {
...
     const fullName = [first, last].filter(Boolean).join(" ");
-    addPoolEntry({
+    const entry: OnlinePoolEntry = {
       fullName,
       firstName: first || undefined,
       lastName: last || undefined,
       otherInfo: poolForm.value.otherInfo.trim() || undefined,
-    });
+    };
     poolForm.value = { firstName: "", lastName: "", otherInfo: "" };
```
[the remark this answers is not in the document]
Comment (glittle) [reply frontend/src/composables/useVoterBallot.ts]: grokbot: Still blocks merge on `32ab223`. `takePoolFormEntry` still clears `poolForm` before `addEntryToNextEmptyVote`; on a full ballot the names are discarded and the voter only sees “ballot full”. Clear only after a successful placement (or check capacity first).

**Review by glittle — COMMENTED** [review]
grokbot: Re-review of #321 at `32ab223` (demotion fix push).

**Fixed / resolved**
- Server never demotes Submitted→Draft (`StatusAfterWrite` + filtered ExecuteUpdate)
- FE autosave sends `isDraft: false` once Submitted
- Tests for Draft excluded from Accept-all and Submitted+IsDraft stays pending

**Still blocks merge**
1. Empty ballot skips autosave (`!hasAnyVote` early return) — clearing the last name leaves the prior Draft on the server; leave/return restores removed names.
2. `takePoolFormEntry` clears the pool form before a free ballot line is confirmed — full ballot discards the entered name.

Imported reserved-location parity still looks good. Keeping draft until those two are fixed.

**Review by glittle — COMMENTED** [review]
grokbot: Re-review of #321 at `5e99710` (empty-overwrite / pool placement push).

**Fixed since last review**
- Pool form: `placePoolFormOnBallot` clears only after a successful placement; full ballot keeps the entered name (thread resolved).
- FE empty-ballot path: `shouldWriteAutosave` / `hasPersistedPayload` correctly *attempts* overwrite; debounce cancel/flush present; demotion guards still look good.
- Accept-all tests for Draft exclusion + Submitted+IsDraft stay pending remain.

**Still blocks merge**
1. `SubmitOnlineBallotDtoValidator` still has `RuleFor(x => x.Votes).NotEmpty()` (unchanged; FluentValidation auto-validation is on). Clearing the last name sends `votes: []` — API 400 — so leave/return still restores removed names. FE Vitest mocks the store and misses this. Allow empty `Votes` for draft overwrite (and for Submitted payload-only clear with `isDraft: false`), plus an integration/unit test through the validator or service.
2. CDN import route title still switched to `nav.importBallots` (English-only); other locales keep `ballots.cdnImport.title`. Non-English headers regress to English fallback.

**Non-blocking**
- Successful Draft autosave sets `isEditing = true`, so the first real Submit can show Update Ballot / updated successfully.
- Monitor `TotalOnlineBallots` counts every row including Draft (Submitted/pending counts stay correct).

Imported reserved-location parity still looks consistent. Keeping draft until (1) is fixed.

File: frontend/src/pages/voting/VoterBallotPage.vue [hunk 6]
```diff
@@ -118,19 +134,73 @@ async function loadElectionData() {
...
+}
+
+const runAutosave = debounce(async () => {
+  if (!autosaveReady.value || !canChangeVote.value || submitting.value) {
+    return;
+  }
+  if (
+    !shouldWriteAutosave(hasAnyVote(votes.value), hasPersistedPayload.value)
+  ) {
+    return;
```
Comment (glittle) [inline frontend/src/pages/voting/VoterBallotPage.vue]: grokbot: **Blocks merge — empty overwrite never reaches the DB.**

`shouldWriteAutosave(false, true)` correctly schedules a silent submit with `votes: []`, but `SubmitOnlineBallotDtoValidator` still has `RuleFor(x => x.Votes).NotEmpty()` and FluentValidation auto-validation is enabled in `Program.cs`. The API returns 400; this catch only logs; leave/return restores the prior Draft names.

Same failure for clearing a Submitted ballot (`isDraft: false`, empty votes).

Relax NotEmpty for draft overwrite (and for Submitted payload-only clear), and add a test that goes through validation/service — the Vitest store mock does not catch this.

File: frontend/src/router/router.ts [hunk 7]
```diff
@@ -234,7 +234,7 @@ const routes = [
 
         component: () => import("../pages/ballots/CdnBallotImportPage.vue"),
 
-        meta: { titleKey: "ballots.cdnImport.title" },
+        meta: { titleKey: "nav.importBallots" },
```
Comment (glittle) [inline frontend/src/router/router.ts]: grokbot: **Blocks merge — i18n regression on CDN import page title.**

`main` used `ballots.cdnImport.title` (present in every locale). This switch to `nav.importBallots` is English-only; fa/fr/es/… have no key, so non-English users get the English fallback in the header.

Keep the route on `ballots.cdnImport.title` (or add `nav.importBallots` to every locale file).

File: frontend/src/pages/voting/VoterBallotPage.vue [hunk 8]
```diff
@@ -118,19 +134,73 @@ async function loadElectionData() {
...
+  }
+
+  try {
+    await onlineVotingStore.submitBallot(
+      electionGuid.value,
+      buildSubmitPayload(autosaveAsDraft(alreadySubmitted.value)),
+      { silent: true },
+    );
+    hasPersistedPayload.value = true;
+    isEditing.value = true;
```
Comment (glittle) [inline frontend/src/pages/voting/VoterBallotPage.vue]: grokbot: Non-blocking: first successful Draft autosave sets `isEditing`, so the initial Submit Ballot path can show Update Ballot / “updated successfully”. Prefer tying editing state to a pre-existing Submitted ballot (`alreadySubmitted` / `whenSubmitted`), not to a Draft write.

**Review by glittle — COMMENTED** [review]
grokbot: Inline notes for `5e99710` — empty Votes fail FluentValidation (blocks leave/return clear); CDN import title i18n regression; isEditing-on-Draft is non-blocking.

File: frontend/src/pages/voting/VoterBallotPage.vue (hunk 4 above)
[the remark this answers is not in the document]
Comment (glittle) [reply frontend/src/pages/voting/VoterBallotPage.vue]: grokbot: On `5e99710` the FE now calls autosave when clearing the last name (`shouldWriteAutosave` / `hasPersistedPayload`). That part is fixed. Leaving this thread open because `SubmitOnlineBallotDtoValidator` still rejects empty `Votes`, so the overwrite never persists — see the new inline on `runAutosave`.

File: backend/Validators/SubmitOnlineBallotDtoValidator.cs [hunk 9]
```diff
@@ -23,9 +23,10 @@ public SubmitOnlineBallotDtoValidator()
             .MaximumLength(250)
             .WithMessage("Voter ID cannot exceed 250 characters");
 
+        // Empty Votes is a payload overwrite (clear last name on Draft or
+        // Submitted). A first visit does not POST an empty ballot.
         RuleFor(x => x.Votes)
-            .NotEmpty()
-            .WithMessage("At least one vote is required");
+            .NotNull();
```
Comment (glittle) [inline backend/Validators/SubmitOnlineBallotDtoValidator.cs]: grokbot: **Blocks merge — empty Submitted + Accept-all locks the voter out with no ballot.**

Allowing `Votes: []` on a non-draft write is correct for leave/return clear, but Status stays `Submitted`, so Accept-all still claims the row.

In `OnlineVotingService.Accept.cs` `ProcessIfStillProcessingAsync`, `hasPendingPayload` requires `payload.Votes.Count > 0`. With an empty overwrite there is no legacy `BallotGuid` either, so Accept-all skips ballot creation and still finalizes as `Processed` (wipes `ListPool`). The voter then cannot change (`CannotChangeOnlineVote`) and no regular ballot exists — silent vote loss if a teller Accept-alls while the voter cleared names mid-edit.

This path was unreachable when FluentValidation required non-empty Votes; this commit opens it. The unit test below also locks in `PendingCount == 1` for empty Submitted.

Fix (prefer without reopening Submitted→Draft demotion): in Accept-all, do not finalize empty non-legacy rows as Processed — restore `Submitted` (or exclude empty payloads from the claim set) and treat as skipped; add a test that Accept-all of empty Submitted does not Process and leaves the voter editable. Draft empty overwrite remains fine (Draft is not pending).

File: Backend.Tests/UnitTests/Services/OnlineVotingServiceAcceptAllTests.cs [hunk 10]
```diff
@@ -42,6 +42,87 @@ public OnlineVotingServiceAcceptAllTests()
...
+
+        var submittedStatus = await _service.GetVoteStatusAsync(
+            election.ElectionGuid, submittedVoter.Email);
+        Assert.True(submittedStatus.HasVoted);
+        Assert.Empty(submittedStatus.PriorVotes);
+        Assert.Equal(OnlineBallotStatus.Submitted, Context.OnlineVotingInfos
+            .Single(o => o.PersonGuid == submittedVoter.Person.PersonGuid).Status);
+
+        var summary = await _service.GetAcceptAllSummaryAsync(election.ElectionGuid);
+        Assert.Equal(1, summary!.PendingCount);
```
Comment (glittle) [inline Backend.Tests/UnitTests/Services/OnlineVotingServiceAcceptAllTests.cs]: grokbot: **Blocks merge — test encodes the Accept-all hazard.**

After empty overwrite, this asserts the row stays `Submitted` and `PendingCount == 1`. That is exactly the row Accept-all will claim and finalize as `Processed` with no ballot (see companion note on the validator / `ProcessIfStillProcessingAsync`).

Replace or extend with coverage that Accept-all on empty Submitted does **not** Process the voter out, and that Draft empty overwrite still stays out of pending.

**Review by glittle — COMMENTED** [review]
grokbot: Re-review of #321 at `41ebc57` (empty-vote validation / CDN title / isEditing push).

**Fixed since last review**
- FluentValidation allows empty `Votes` overwrite (`NotNull` not `NotEmpty`); FE+integration tests cover Draft/Submitted clear
- CDN import route title restored to `ballots.cdnImport.title`
- `isEditing` follows Submitted (`whenSubmitted` / explicit submit), not Draft autosave
- Prior demotion / pool-form / empty-FE-autosave threads remain resolved

**Still blocks merge**
1. Empty `Submitted` stays Accept-all pending; Accept-all finalizes `Processed` with no regular ballot when `Votes` is empty — silent vote loss mid-edit. Do not demote to Draft; skip/restore empty non-legacy rows in Accept-all and add a test.

Keeping draft. Do not merge.

File: backend/Validators/SubmitOnlineBallotDtoValidator.cs (hunk 9 above)
Comment (glittle) [reply backend/Validators/SubmitOnlineBallotDtoValidator.cs]: grokbot: Confirmed on `6994bc76`. Empty Submitted is left out of the Accept-all claim set (`HasVotesToAccept`); a leftover empty Processing claim is restored to Submitted (never Draft / Processed). Unit + concurrency tests cover Accept-all not Processing and voter stays editable. Resolving.

File: Backend.Tests/UnitTests/Services/OnlineVotingServiceAcceptAllTests.cs (hunk 10 above)
Comment (glittle) [reply Backend.Tests/UnitTests/Services/OnlineVotingServiceAcceptAllTests.cs]: grokbot: Confirmed on `6994bc76`. `Submit_EmptyVotes_OverwritesDraftAndSubmittedPayload` now asserts Accept-all AcceptedCount 0 / no ballot / still CanChangeVote, plus `AcceptAll_EmptyProcessingClaim_RestoresSubmittedAndLeavesVoterEditable` and the SQLite concurrency twin. Resolving.
